File: examples/dotnet/AccessTokenGenerator.cs
using System.IdentityModel.Tokens.Jwt;
using System.Net.Http.Json;
using System.Security.Claims;
using System.Security.Cryptography;
using Microsoft.IdentityModel.Tokens;

namespace dotnet
{
  public class AccessTokenGenerator
  {
    public static async Task<string> Generate(string identityProviderUrl, string projectIdentifier, PrivateApiKey privateApiKey)
    {
      try
      {
        RSA rsa = RSA.Create();

        rsa.ImportFromPem(privateApiKey.key);

        RsaSecurityKey rsaSecurityKey = new(rsa)
        {
          KeyId = privateApiKey.keyId,
        };

        SigningCredentials signingCredentials = new(rsaSecurityKey, SecurityAlgorithms.RsaSha256);

        JwtSecurityToken jwtSecurityToken = new(
          claims: [
            new Claim("iat", DateTimeOffset.UtcNow.ToUnixTimeSeconds().ToString(), ClaimValueTypes.Integer),
            new Claim("iss", privateApiKey.userId),
            new Claim("sub", privateApiKey.userId),
            new Claim("aud", identityProviderUrl),
            new Claim("exp", ((DateTimeOffset)DateTime.UtcNow.AddMinutes(1)).ToUnixTimeSeconds().ToString(), ClaimValueTypes.Integer)
          ],
          signingCredentials: signingCredentials
        );

        string jwtSignedToken = new JwtSecurityTokenHandler().WriteToken(jwtSecurityToken);

        Dictionary<string, string> requestParameters = new()
        {
          { "grant_type", "urn:ietf:params:oauth:grant-type:jwt-bearer" },
          { "assertion", jwtSignedToken },
          { "scope", $"openid profile urn:zitadel:iam:org:project:id:{projectIdentifier}:aud" }
        };

        HttpClient httpClient = new()
        {
          BaseAddress = new Uri(identityProviderUrl),
          Timeout = TimeSpan.FromSeconds(3)
        };

        // For the API to accept requests we need to set the `User-Agent` header parameter
        httpClient.DefaultRequestHeaders.Add("User-Agent", "AccessTokenGenerator/1.0");

        TokenResponse response = await httpClient.PostAsync(
            "/oauth/v2/token",
            new FormUrlEncodedContent(requestParameters)
          )
          .Result
          .Content
          .ReadFromJsonAsync<TokenResponse>();

        return response.access_token;
      }
      catch (Exception exception)
      {
        throw new Exception("Failed to generate access token", exception);
      }
    }

    private struct TokenResponse
    {
      public required string access_token { get; set; }
    }
  }
}

File: examples/python/access_token_generator.py
import httpx
import time
import jwt
from cryptography.hazmat.backends import default_backend
from cryptography.hazmat.primitives import serialization
from typing import Dict
from data_structures import PrivateApiKey


class AccessTokenGenerator:

    @staticmethod
    def generate(
        identity_provider_url: str,
        project_identifier: str,
        private_api_key: PrivateApiKey,
    ) -> str:
        try:
            private_key = serialization.load_pem_private_key(
                private_api_key.key.encode(), password=None, backend=default_backend()
            )

            headers = {"kid": private_api_key.key_id}

            current_time = int(time.time())

            claims = {
                "iat": current_time,
                "iss": private_api_key.user_id,
                "sub": private_api_key.user_id,
                "aud": identity_provider_url,
                "exp": current_time + 60,
            }

            jwt_signed_token = jwt.encode(
                claims, private_key, algorithm="RS256", headers=headers
            )

            request_parameters: Dict[str, str] = {
                "grant_type": "urn:ietf:params:oauth:grant-type:jwt-bearer",
                "assertion": jwt_signed_token,
                "scope": f"openid profile urn:zitadel:iam:org:project:id:{project_identifier}:aud",
            }

            httpClient = httpx.Client(base_url=identity_provider_url, timeout=3)

            response = httpClient.post("/oauth/v2/token", data=request_parameters)
            response.raise_for_status()
            responseAsJson = response.json()

            return responseAsJson["access_token"]
        except Exception as exception:
            raise Exception("Failed to generate access token") from exception
